src/Microsoft.OData.Core/UriParser/Parsers/CountSegmentParser.cs

```diff
@@ -0,0 +1,138 @@
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.UriParser
+{
+    using System;
+    using ODataErrorStrings = Microsoft.OData.Strings;
+
+    /// <summary>
+    /// Syntactic parser for the $count segment.
+    /// </summary>
+    internal sealed class CountSegmentParser
+    {
+        /// <summary>
+        /// Reference to the lexer.
+        /// </summary>
+        private readonly ExpressionLexer lexer;
+
+        /// <summary>
+        /// Method used to parse arguments.
+        /// </summary>
+        private readonly UriQueryExpressionParser parser;
+
+        /// <summary>
+        /// Create a new CountSegmentParser.
+        /// </summary>
+        /// <param name="lexer">Lexer positioned at a $count identifier.</param>
+        /// <param name="parser">The UriQueryExpressionParser.</param>
+        public CountSegmentParser(ExpressionLexer lexer, UriQueryExpressionParser parser)
+        {
+            ExceptionUtils.CheckArgumentNotNull(lexer, "lexer");
+            ExceptionUtils.CheckArgumentNotNull(parser, "parser");
+            this.lexer = lexer;
+            this.parser = parser;
+        }
+
+        /// <summary>
+        /// Reference to the underlying UriQueryExpressionParser.
+        /// </summary>
+        public UriQueryExpressionParser UriQueryExpressionParser
+        {
+            get { return this.parser; }
+        }
+
+        /// <summary>
+        /// Reference to the lexer.
+        /// </summary>
+        public ExpressionLexer Lexer
+        {
+            get { return this.lexer; }
+        }
+
+        /// <summary>
+        /// Create a <see cref="CountSegmentToken"/>.
+        /// </summary>
+        /// <param name="identifier">The name of this token, which in this case is always "$count".The name of this token, which in this case is always "$count".</param>
+        /// <param name="countedInstance">The instance to count on.</param>
+        public CountSegmentToken CreateCountSegmentToken(string identifier, QueryToken countedInstance)
+        {
+            QueryToken filterToken = null;
+            QueryToken searchToken = null;
+            string textWithinParenthesis = null;
+
+            // We need to maintain the recursionDepth of the outer $filter query since calling
+            // ParseFilter or ParseSearch below will reset it to 0.
+            int outerRecursiveDepth = this.UriQueryExpressionParser.recursionDepth;
+            ExpressionLexer outerLexer = this.UriQueryExpressionParser.Lexer;
+
+            if (this.lexer.CurrentToken.Kind != ExpressionTokenKind.OpenParen)
+            {
+                return new CountSegmentToken(identifier, countedInstance);
+            }
+            else
+            {
+                textWithinParenthesis = this.lexer.AdvanceThroughExpandOption();

```

Comment:
>AdvanceThroughExpandOption [](http://example.com/codeflow?start=51&length=26)

caution -- you are using AdvanceThroughExpandOption differently than how it is used in selectexpandbinder.  There it is used to read a single option value only (i.e., after the equals sign). Here you using it to read the entire option key/value pair. Special logic within AdvanceThroughExpandOption for handling things like quoting (now or in the future) might make assumptions that it is only reading the value.  Better might be to first parse the query option name and then use AdvanceThroughExpandOption to parse just the value of the query option (i.e., everything after the equals).

Reply:
Resolved